subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cntr.c

```diff
@@ -47,17 +116,89 @@ uint32_t cntr_stop(void)
 		return 1;
 	}
 
+#if defined(CONFIG_BT_CTLR_NRF_GRTC)
+	/* TODO: if we own and stop GRTC, implement stop here */
+#else /* !CONFIG_BT_CTLR_NRF_GRTC */
 	nrf_rtc_task_trigger(NRF_RTC, NRF_RTC_TASK_STOP);
+#endif /* !CONFIG_BT_CTLR_NRF_GRTC */
 
 	return 0;
 }
 
 uint32_t cntr_cnt_get(void)
 {
+#if defined(CONFIG_BT_CTLR_NRF_GRTC)
+	uint32_t cntr_l, cntr_h, cntr_h_overflow;
+
+	/* NOTE: For a 32-bit implementation, L value is read after H
+	 *       to avoid another L value after SYSCOUNTER gets ready.
+	 *       If both H and L values are desired, then swap the order and
+	 *       ensure that L value does not change when H value is read.
+	 */
+	do {
+		cntr_h = nrf_grtc_sys_counter_high_get(NRF_GRTC);
+		cntr_l = nrf_grtc_sys_counter_low_get(NRF_GRTC);
+		cntr_h_overflow = nrf_grtc_sys_counter_high_get(NRF_GRTC);
+	} while ((cntr_h & GRTC_SYSCOUNTER_SYSCOUNTERH_BUSY_Msk) ||
+		 (cntr_h_overflow & GRTC_SYSCOUNTER_SYSCOUNTERH_OVERFLOW_Msk));
+
+	return cntr_l;
+#else /* !CONFIG_BT_CTLR_NRF_GRTC */
 	return nrf_rtc_counter_get(NRF_RTC);
+#endif /* !CONFIG_BT_CTLR_NRF_GRTC */
 }
 
 void cntr_cmp_set(uint8_t cmp, uint32_t value)
 {
+#if defined(CONFIG_BT_CTLR_NRF_GRTC)
+	uint32_t cntr_l, cntr_h, cntr_h_overflow, stale;
+
+	/* NOTE: We are going to use TASKS_CAPTURE to read current
+	 *       SYSCOUNTER H and L, so that COMPARE registers can be set
+	 *       considering that we need to set H compare value too.
+	 */
+
+	/* Read current syscounter value */
+	do {
+		cntr_h = nrf_grtc_sys_counter_high_get(NRF_GRTC);
+		cntr_l = nrf_grtc_sys_counter_low_get(NRF_GRTC);
+		cntr_h_overflow = nrf_grtc_sys_counter_high_get(NRF_GRTC);
+	} while ((cntr_h & GRTC_SYSCOUNTER_SYSCOUNTERH_BUSY_Msk) ||
+		 (cntr_h_overflow & GRTC_SYSCOUNTER_SYSCOUNTERH_OVERFLOW_Msk));
+
+	/* Disable capture/compare */
+	nrf_grtc_sys_counter_compare_event_disable(NRF_GRTC, cmp);
+
+	/* Set a stale value in capture value */
+	stale = cntr_l - 1U;
+	NRF_GRTC->CC[cmp].CCL = stale;
+
+	/* Trigger a capture */
+	nrf_grtc_task_trigger(NRF_GRTC, (NRF_GRTC_TASK_CAPTURE_0 + (cmp * sizeof(uint32_t))));
+
+	/* Wait to get a new L value */
+	do {
+		cntr_l = NRF_GRTC->CC[cmp].CCL;
+	} while (cntr_l == stale);
+
+	/* Read H value */
+	cntr_h = NRF_GRTC->CC[cmp].CCH;
+
+	/* NOTE: HERE, we have cntr_h and cntr_l in sync. */
+
+	/* Handle rollover between current and expected value */
+	if (value < cntr_l) {
+		cntr_h++;
+	}
+
+	/* Set compare register values */
+	nrf_grtc_sys_counter_cc_set(NRF_GRTC, cmp,
+				    ((((uint64_t)cntr_h & GRTC_CC_CCH_CCH_Msk) << 32) | value));
+
+	/* Enable compare */
+	nrf_grtc_sys_counter_compare_event_enable(NRF_GRTC, cmp);
```

Comment:
This would be redundant.
The CC channel should have been already enabled when doing nrf_grtc_sys_counter_cc_set() (as a sideeffect of writing to CCH )

Reply:
My understanding from the docs is that CC[cc_channel].CCEN is automatically set when we write to CC[cc_channel].CCH

Reply:
Discussed offline that cleaning this up can be done in a follow up